packages/hdwallet-xdefi/src/adapter.ts

```diff
@@ -0,0 +1,42 @@
+import * as core from "@shapeshiftoss/hdwallet-core";
+import { XDeFiHDWallet } from "./xdefi";
+
+export class XDeFiAdapter {
+  keyring: core.Keyring;
+
+  // wallet id to remove from the keyring when the active wallet changes
+  currentDeviceID?: string;
+
+  private constructor(keyring: core.Keyring) {
+    this.keyring = keyring;
+  }
+
+  public static useKeyring(keyring: core.Keyring) {
+    return new XDeFiAdapter(keyring);
+  }
+
+  public async initialize(): Promise<number> {
+    return Object.keys(this.keyring.wallets).length;
+  }
+
+  public async pairDevice(): Promise<core.HDWallet> {
+    const provider: any = (window as any).xfi && (window as any).xfi.ethereum;
+    if (!provider) {
+      window.open("https://xdefi.io", "_blank");
+    }
+    try {
+      await provider.request({ method: "eth_requestAccounts" });
+    } catch (error) {
+      console.error("Could not get XDeFi accounts. ");
+      throw error;
+    }
+    const wallet = new XDeFiHDWallet();
+    await wallet.initialize();
```

Comment:
Just noticed that `provider` doesn't actually get passed into the instance here!

Reply:
The `provider` is not passed as that's how it the method signature is defined in `code.HDWallet`. Should the provider be passed into the instance instead?

```
  /**
   * Initialize a device session.
   */
  initialize(): Promise<any>;
  ```

Reply:
Ah, gotcha. We can still extend that interface, we just have to do it a bit wonkily -- see the matching comment on the wallet implementation for details.
```suggestion
    const wallet = new XDeFiHDWallet();
    await wallet.initialize(provider);
```